pkg/pillar/uuidpairtonum/uuidpairtonum.go

```diff
@@ -0,0 +1,143 @@
+	baseID uuid.UUID, appID uuid.UUID, appNum int, mustCreate bool,
+	numType string) {
+	log.Functionf("NumAllocate(%s, %s, %d, %v)", baseID.String(),
+		appID.String(), appNum, mustCreate)
+	now := time.Now()
+	key := types.UUIDPairToNumKey(baseID, appID)
+	i, err := pub.Get(key)
+	if err != nil {
+		u := types.UUIDPairToNum{
+			BaseID:      baseID,
+			AppID:       appID,
+			CreateTime:  now,
+			LastUseTime: now,
+			InUse:       true,
+			NumType:     numType,
+			Number:      appNum,
+		}
+		log.Functionf("NumAllocate(%s) publishing %v",
+			key, u)
+		pub.Publish(u.Key(), u)
+		return
+	}
+	u := i.(types.UUIDPairToNum)
+	if u.NumType != numType {
+		log.Fatalf("NumAllocate(%s) wrong numType %s vs. %s",
+			key, u.NumType, numType)
+	}
+	if mustCreate {
+		log.Fatalf("NumAllocate(%s) already exists %v",
+			key, u)
+	}
+	if u.Number != appNum {
+		log.Warnf("NumAllocate(%s) number changing from %d to %d",
+			key, u.Number, appNum)
+	}
+	if u.InUse {
+		log.Warnf("NumAllocate(%s) already InUse %v",
+			key, u)
+	}
+	u.Number = appNum
+	u.InUse = true
+	u.LastUseTime = time.Now()
+	// XXX note that nothing but lastusetime might be updated! Improve log?
+	log.Functionf("NumAllocate(%s) publishing updated %v",
+		key, u)
+	if err := pub.Publish(u.Key(), u); err != nil {
+		// Could be low on disk space
+		log.Errorf("NumAllocate(%s) publish failed %v",
+			key, err)
+	}
+	return
+}
+
+// NumFree : Clears InUse flag
+func NumFree(log *base.LogObject, pub pubsub.Publication,
+	baseID uuid.UUID, appID uuid.UUID) {
+	key := types.UUIDPairToNumKey(baseID, appID)
+	i, err := pub.Get(key)
+	if err != nil {
+		log.Fatalf("NumFree(%s) does not exist", key)
+	}
+	u := i.(types.UUIDPairToNum)
+	u.InUse = false
+	u.LastUseTime = time.Now()
+	log.Functionf("NumFree(%s) publishing updated %v",
+		key, u)
+	if err := pub.Publish(u.Key(), u); err != nil {
+		// Could be low on disk space
+		log.Errorf("NumFree(%s) publish failed %v",
+			key, err)
+	}
+}
+
+// NumDelete : Removes the integer map for a given UUID pair
+func NumDelete(log *base.LogObject, pub pubsub.Publication,
+	baseID uuid.UUID, appID uuid.UUID) {
+	key := types.UUIDPairToNumKey(baseID, appID)
+	_, err := pub.Get(key)
+	if err != nil {
+		log.Fatalf("NumDelete(%s) does not exist", key)
+	}
+	if err := pub.Unpublish(key); err != nil {
+		log.Fatalf("NumDelete(%s) unpublish failed %v", key, err)
+	}
+}
+
+// NumGetOldestUnused : returns a number, not recently in use
+func NumGetOldestUnused(log *base.LogObject, pub pubsub.Publication,
+	baseID uuid.UUID, numType string) (uuid.UUID, int, error) {
+	log.Functionf("NumGetOldestUnused(%s)", numType)
+	// Will have a LastUseTime of zero
+	oldest := new(types.UUIDPairToNum)
+	items := pub.GetAll()
+	for _, st := range items {
+		item := st.(types.UUIDPairToNum)
+		if item.NumType != numType || item.InUse || item.BaseID != baseID {
+			continue
+		}
+		if oldest.LastUseTime.Before(item.LastUseTime) {
```

Comment:
Please check the sample: https://play.golang.org/p/TuJc02sAVay. Current implementation returns not the oldest, but the newest. We should use something like `oldest.LastUseTime.IsZero() || oldest.LastUseTime.After(item.LastUseTime)`.

Reply:
My point is that that bug is unrelated to this PR since this code is a duplicate of the code in uuidtonum.go. So better to do fix that bug as a separate PR since this PR has enough on its plate than fix existing bugs.